### PR TITLE
chore(charts): add ndm helm3 charts

### PR DESCRIPTION
**Pull request by sonasingh46**
**NOTE:** This PR should be merged only after #491 and #490 
This PR adds helm3 charts for NDM installation.

NDM  component will have its own service account and RBAC. So following Kubernetes resource are 
created for NDM that is not shared with other OpenEBS components:
1. ClusterRole: ( Default name is `openebs-ndm`)
2. ClusterRoleBinding: ( Default name is `openebs-ndm`)
3. ServiceAccount : ( Default name is `openebs-ndm`)

NDM Daemonset default name : `openebs-ndm`
NDM deployment default name: `openebs-ndm-operator`
NDM configmap default name:  `openebs-ndm-config`

#### Special notes for your reviewer:

- There has been a change in match label selectors for the NDM pods. See the following for the comparison 

**Older Helm Chart Based NDM Daemon Set:** 

```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: {{ template "openebs.fullname" . }}-ndm
  labels:
    app: {{ template "openebs.name" . }}
    chart: {{ template "openebs.chart" . }}
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
    component: ndm
    openebs.io/component-name: ndm
    openebs.io/version: {{ .Values.release.version }}
spec:
  updateStrategy:
    type: "RollingUpdate"
  selector:
    matchLabels:
      app: {{ template "openebs.name" . }}
      release: {{ .Release.Name }}
      component: ndm
  template:
    metadata:
      labels:
        app: {{ template "openebs.name" . }}
        release: {{ .Release.Name }}
        component: ndm
        openebs.io/component-name: ndm
        name: openebs-ndm
        openebs.io/version: {{ .Values.release.version }}
```

**Current Helm Chart Based NDM Daemon Set:** 

```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  annotations:
    deprecated.daemonset.template.generation: "1"
    meta.helm.sh/release-name: pascal
    meta.helm.sh/release-namespace: openebs
  creationTimestamp: "2020-08-26T15:28:33Z"
  generation: 1
  labels:
    app: openebs-ndm
    app.kubernetes.io/managed-by: Helm
    chart: openebs-ndm-0.8.0
    component: ndm
    heritage: Helm
    openebs.io/component-name: ndm
    openebs.io/version: 0.8.0
    release: pascal
spec:
  selector:
    matchLabels:
      app: openebs-ndm
      component: ndm
      release: pascal
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: openebs-ndm
        chart: openebs-ndm-0.8.0
        component: ndm
        heritage: Helm
        name: openebs-ndm
        openebs.io/component-name: ndm
        openebs.io/version: 0.8.0
        release: pascal
```

**Older Helm Chart Based NDM Operator:** 

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{ template "openebs.fullname" . }}-ndm-operator
  labels:
    app: {{ template "openebs.name" . }}
    chart: {{ template "openebs.chart" . }}
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
    component: ndm-operator
    openebs.io/component-name: ndm-operator
    openebs.io/version: {{ .Values.release.version }}
    name: ndm-operator
spec:
  replicas: {{ .Values.ndmOperator.replicas }}
  strategy:
    type: "Recreate"
    rollingUpdate: null
  selector:
    matchLabels:
      app: {{ template "openebs.name" . }}
      release: {{ .Release.Name }}
  template:
    metadata:
      labels:
        app: {{ template "openebs.name" . }}
        release: {{ .Release.Name }}
        component: ndm-operator
        name: ndm-operator
        openebs.io/component-name: ndm-operator
        openebs.io/version: {{ .Values.release.version }}
```

**Current Helm Chart Based NDM operator:**

```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: pascal
    meta.helm.sh/release-namespace: openebs
  creationTimestamp: "2020-08-26T15:28:33Z"
  generation: 1
  labels:
    app: openebs-ndm-operator
    app.kubernetes.io/managed-by: Helm
    chart: openebs-ndm-0.8.0
    component: openebs-ndm-operator
    heritage: Helm
    openebs.io/component-name: openebs-ndm-operator
    openebs.io/version: 0.8.0
    release: pascal
spec:
  replicas: 1
  selector:
    matchLabels:
      app: openebs-ndm-operator
      component: openebs-ndm-operator
      release: pascal
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: openebs-ndm-operator
        chart: openebs-ndm-0.8.0
        component: openebs-ndm-operator
        heritage: Helm
        openebs.io/component-name: openebs-ndm-operator
        openebs.io/version: 0.8.0
        release: pascal

``` 

**Notable Changes As Part of Best Practice:**
- The labels of ndm daemonset and ndm operator yaml comes from `helper.tpl`. This is more maintainable and less error prone.
- Clean up container image values is segregated in Values.yaml file.

### CODE REVIEW

File: helm/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+- Assign openebs namespace to the current context:
+```bash
+kubectl config set-context <current_context_name> --namespace=openebs
+```
+
+- Add NDM helm repo:
+```bash
+helm repo add openebs https://openebs.github.io/charts/ndm
```
Comment (akhilerm) [inline helm/README.md:26]: Where does this URL point to?

Comment (sonasingh46) [reply helm/README.md:26]: URL is wrong... I need to update it once github actions and gh-pages are set.

Comment (sonasingh46) [reply helm/README.md:26]: I will fix this and update... If this PR needs to still get merged before the fixes -- I will remove this wrong URL

**Comment by akhilerm** [comment]
@sonasingh46 Is this PR ready to be reviewed?

**Comment by sonasingh46** [comment]
@akhilerm -- Yes

File: charts/ndm/values.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  probes:
+    enableSeachest: true
+    enableUdevProbe: true
+    enableSmartProbe: true
+  healthCheck:
+    initialDelaySeconds: 30
+    periodSeconds: 60
+
+ndmOperator:
+  name: openebs-ndm-operator
```
Comment (kmova) [inline charts/ndm/values.yaml:46]: why is this name `openebs-ndm-operator` and not just `ndm-operator` like in above?

**Comment by akhilerm** [comment]
Changes in labels and naming as per the above chart (from the PR description) compared with openebs/charts

**1. NDM Daemonset**
```diff
apiVersion: apps/v1
kind: DaemonSet
metadata:
  annotations:
    deprecated.daemonset.template.generation: "1"
    meta.helm.sh/release-name: pascal
    meta.helm.sh/release-namespace: openebs
  creationTimestamp: "2020-08-26T15:28:33Z"
  generation: 1
  labels:
-   app: openebs
+   app: openebs-ndm
    app.kubernetes.io/managed-by: Helm
-   chart: openebs-2.2.0
+   chart: openebs-ndm-0.8.0
    component: ndm
    heritage: Helm
    openebs.io/component-name: ndm
-   openebs.io/version: 2.2.0
+   openebs.io/version: 0.8.0
    release: pascal
spec:
  selector:
    matchLabels:
-     app: openebs
+     app: openebs-ndm
      component: ndm
      release: pascal
  template:
    metadata:
      creationTimestamp: null
      labels:
-       app: openebs
+       app: openebs-ndm
+       chart: openebs-ndm-0.8.0
        component: ndm
+       heritage: Helm
        name: openebs-ndm
        openebs.io/component-name: ndm
-       openebs.io/version: 2.2.0
+       openebs.io/version: 0.8.0
        release: pascal
```

**2. NDM Operator**
```diff
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: pascal
    meta.helm.sh/release-namespace: openebs
  creationTimestamp: "2020-08-26T15:28:33Z"
  generation: 1
  labels:
-   app: openebs
+   app: openebs-ndm-operator
    app.kubernetes.io/managed-by: Helm
-   chart: openebs-2.2.0
+   chart: openebs-ndm-0.8.0
-   component: ndm-operator
+   component: openebs-ndm-operator
    heritage: Helm
-   openebs.io/component-name: ndm-operator
-   openebs.io/version: 2.2.0
+   openebs.io/component-name: openebs-ndm-operator
+   openebs.io/version: 0.8.0
    release: pascal
spec:
  replicas: 1
  selector:
    matchLabels:
-     app: openebs
+     app: openebs-ndm-operator
+     component: openebs-ndm-operator
      release: pascal
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
-       app: openebs
+       app: openebs-ndm-operator
+       chart: openebs-ndm-0.8.0
-       component: ndm-operator
+       component: openebs-ndm-operator
+       heritage: Helm
-       openebs.io/component-name: ndm-operator
-       openebs.io/version: 2.2.0
+       openebs.io/component-name: openebs-ndm-operator
+       openebs.io/version: 0.8.0
        release: pascal
```

@sonasingh46 @kmova , This is the comparison with the current charts from openebs/charts and the charts in this PR.

Have some queries: 
1. On the `.spec.template.labels`. The `openebs.io/component-name` label has the `openebs` prefix for NDM operator but not for NDM daemon set.

File: charts/ndm/values.yaml (hunk 2 above)
Comment (sonasingh46) [reply charts/ndm/values.yaml]: fixed it.

**Comment by sonasingh46** [comment]
@akhilerm -- 
I am not sure why only ndm is there.
Do we want the label in operator as `ndm-operator` ? If yes -- it can be done by setting it in the values.yaml. 
Same thing is happening for the ndm daemonset.  If label is skipped -- this full thing will come `openebs-ndm-operator`

**Comment by sonasingh46** [comment]
@akhilerm @kmova -- Please let me know further action items on the label part and in general on this PR.

File: deploy/charts/templates/daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,162 @@
+{{- if .Values.ndm.enabled }}
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: {{ template "openebs-ndm.fullname" . }}
+  {{- if .Values.ndm.annotations }}
```
Comment (davidkarlsen) [inline deploy/charts/templates/daemonset.yaml:6]: This can be shortened to make it more readable like this:

```suggestion
  {{- with .Values.ndm.annotations }}
  annotations: {{ toYaml . | nindent 4 }}
  {{- end }}
```
See https://banzaicloud.com/blog/creating-helm-charts-part-2/

same goes for other places.

File: deploy/charts/Chart.yaml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 apiVersion: v2
 name: openebs-ndm
 description: Kubernetes Storage Device Management. Also used to maintain the inventory of block devices.
-version: 0.8.0
-appVersion: 0.8.0
+version: 0.8.1
+appVersion: 0.8.1
```
Comment (akhilerm) [inline deploy/charts/Chart.yaml:5]: current version is 0.9.0

File: deploy/charts/templates/_helpers.tpl [hunk 5]
```diff
@@ -106,14 +111,14 @@ Create match labels for ndm operator deployment
...
-component: {{ .Values.ndmOperator.name | quote }}
+component: {{ default (include "openebs-ndm.operator.name" .) .Values.ndmOperator.componentName }}
 {{- end -}}
 
 {{/*
 Create component labels for ndm operator component
 */}}
 {{- define "openebs-ndm.operator.componentLabels" -}}
-openebs.io/component-name: {{ .Values.ndmOperator.name | quote }}
+openbes.io/component-name: {{ default (include "openebs-ndm.operator.name" .) .Values.ndmOperator.componentName }}
```
Comment (akhilerm) [inline deploy/charts/templates/_helpers.tpl:121]:
```suggestion
openebs.io/component-name: {{ default (include "openebs-ndm.operator.name" .) .Values.ndmOperator.componentName }}
```

File: deploy/charts/values.yaml [hunk 6]
```diff
@@ -11,9 +11,9 @@ ndm:
     # For example : quay.io/ is a correct value here and quay.io is incorrect
     registry:
     repository: openebs/node-disk-manager
-    pullPolicy: IfNotPresent
+    pullPolicy: Always
```
Comment (akhilerm) [inline deploy/charts/values.yaml:14]: It should be `IfNotPresent` as we are always using tagged images.

File: deploy/charts/values.yaml [hunk 7]
```diff
@@ -11,9 +11,9 @@ ndm:
...
     # Overrides the image tag whose default is the chart appVersion.
-    tag: 0.8.0
+    tag: 0.8.1
```
Comment (akhilerm) [inline deploy/charts/values.yaml:16]: Version need to be 0.9.0

**Comment by akhilerm** [comment]
@sonasingh46 Given some more nitpick comments. Else everything looks good.

File: deploy/charts/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  - block-storage
+  - ndm
+  - disk-inventory
+  - storage
+sources:
+  - https://github.com/openebs/node-disk-manager
+maintainers:
+  - name: akhilerm
+    email: akhil.mohan@mayadata.io
+  - name: kmova
```
Comment (kmova) [inline deploy/charts/Chart.yaml:19]: @sonasingh46 - can you replace my name with Michael ( @xUnholy )

File: deploy/charts/templates/rbac.yaml [hunk 9]
```diff
@@ -0,0 +1,44 @@
+{{- if .Values.serviceAccount.create -}}
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ include "openebs-ndm.serviceAccountName" . }}
+{{- end }}
+---
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (kmova) [inline deploy/charts/templates/rbac.yaml:9]: check if this can be moved to v1. Looks like v1beta1 will be deprecated soon.

**Comment by kmova** [comment]
@sonasingh46 - can we set the name of the directory as "helm" instead of "charts". Charts can refer to other types of charts like rancher, ibm or nks charts etc.

File: deploy/charts/Chart.yaml (hunk 8 above)
Comment (sonasingh46) [reply deploy/charts/Chart.yaml:19]: @xUnholy -- Can you please comment your email address here that I can put in here?

Comment (xUnholy) [reply deploy/charts/Chart.yaml:19]: This email should work with notifications 20387402+xUnholy@users.noreply.github.com

Unless @kmova you'd rather my contactable email?

File: deploy/charts/templates/_helpers.tpl (hunk 5 above)
Comment (sonasingh46) [reply deploy/charts/templates/_helpers.tpl:121]: That has to be in a single line.

Comment (akhilerm) [reply deploy/charts/templates/_helpers.tpl]: Oops. Sorry this is not about singe line / multi line. There was a typo in `openebs`

Comment (sonasingh46) [reply deploy/charts/templates/_helpers.tpl]: Oh. Ok thank you. I missed that.

File: deploy/charts/templates/daemonset.yaml (hunk 3 above)
Comment (sonasingh46) [reply deploy/charts/templates/daemonset.yaml]: done

File: deploy/charts/templates/rbac.yaml (hunk 9 above)
Comment (sonasingh46) [reply deploy/charts/templates/rbac.yaml]: done

**Comment by sonasingh46** [comment]
@davidkarlsen @kmova @akhilerm -- Have incorporated the review comments.

File: deploy/charts/Chart.yaml (hunk 8 above)
Comment (sonasingh46) [reply deploy/charts/Chart.yaml]: Done

File: deploy/helm/values.yaml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    count: "0"
+  updateStrategy:
+    type: RollingUpdate
+  annotations: {}
+  podAnnotations: {}
+  ## Labels to be added to ndm daemonset pods
+  podLabels:
+    name: openebs-ndm
+  nodeSelector: {}
+  tolerations: []
```
Comment (prateekpandey14) [inline deploy/helm/values.yaml:29]: we should add resourcelimits tag as well , to configure the limits for the deployments, Daemonset templates

File: deploy/helm/values.yaml [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  image:
+    registry:
+    repository: openebs/node-disk-operator
+    pullPolicy: IfNotPresent
+    # Overrides the image tag whose default is the chart appVersion.
+    tag: 0.9.0
+  annotations: {}
+  podAnnotations: {}
+  nodeSelector: {}
+  tolerations: []
```
Comment (prateekpandey14) [inline deploy/helm/values.yaml:57]: Same as above comments

File: deploy/helm/values.yaml [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+ndmOperator:
+  name: operator
+  enabled: true
+  image:
+    registry:
+    repository: openebs/node-disk-operator
+    pullPolicy: IfNotPresent
+    # Overrides the image tag whose default is the chart appVersion.
+    tag: 0.9.0
```
Comment (prateekpandey14) [inline deploy/helm/values.yaml:53]: nit: update to the latest release tag ?

File: deploy/helm/values.yaml [hunk 13]
```diff
@@ -0,0 +1,99 @@
+# Default values for ndm.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+release:
+  version: "0.9.0"
```
Comment (prateekpandey14) [inline deploy/helm/values.yaml:5]: nit: update to latest release tag ?

Comment (akhilerm) [reply deploy/helm/values.yaml:5]: latest is now 0.9.1, there was a dot release last week for a bug fix.

File: deploy/helm/values.yaml [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    name: openebs-ndm
+  nodeSelector: {}
+  tolerations: []
+  filters:
+    enableOsDiskExcludeFilter: true
+    enableVendorFilter: true
+    excludeVendors: "CLOUDBYT,OpenEBS"
+    enablePathFilter: true
+    includePaths: ""
+    excludePaths: "loop,fd0,sr0,/dev/ram,/dev/dm-,/dev/md,/dev/rbd"
```
Comment (akhilerm) [inline deploy/helm/values.yaml:36]:
```suggestion
    excludePaths: "loop,fd0,sr0,/dev/ram,/dev/dm-,/dev/md,/dev/rbd,/dev/zd"
```
zfs volumes has also been added to the default exclude list.

File: deploy/helm/templates/daemonset.yaml [hunk 15]
```diff
@@ -157,4 +162,11 @@ spec:
       tolerations:
 {{ toYaml .Values.ndm.tolerations | indent 8 }}
 {{- end }}
+{{- if .Values.ndm.securityContext }}
+      securityContext:
+{{ toYaml .Values.ndm.securityContext | indent 8 }}
+{{- end }}
+{{- if .Values.ndm.hostNetwork }}
+      hostNetwork: true
```
Comment (akhilerm) [inline deploy/helm/templates/daemonset.yaml:170]: @sonasingh46  Why was this change required? This is mandatory for NDM to work. If users disable it, we wont receive any udev events.

**Comment by prateekpandey14** [comment]
@sonasingh46 changes looks good, only one change requested by @akhilerm needs to be resolve.

File: deploy/helm/templates/daemonset.yaml (hunk 15 above)
Comment (sonasingh46) [reply deploy/helm/templates/daemonset.yaml:169]: changed

**Review by akhilerm — APPROVED** [review]
lgtm

**Review by kmova — APPROVED** [review]
